adapters/smilewanted/smilewanted.go

```diff
@@ -0,0 +1,162 @@
+package smilewanted
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/mxmCherry/openrtb/v15/openrtb2"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type SmileWantedAdapter struct {
+	URI string
+}
+
+func (a *SmileWantedAdapter) MakeRequests(requestIn *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+
+	request := *requestIn
+
+	if len(request.Imp) == 0 {
+		return nil, []error{&errortypes.BadInput{
+			Message: "No impression in the bid request",
+		}}
+	}
+
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(request.Imp[0].Ext, &bidderExt); err != nil {
+		return nil, []error{&errortypes.BadInput{
+			Message: "ext.bidder not provided",
+		}}
+	}
+
+	var smilewantedExt openrtb_ext.ExtImpSmileWanted
+	if err := json.Unmarshal(bidderExt.Bidder, &smilewantedExt); err != nil {
+		return nil, []error{&errortypes.BadInput{
+			Message: "ext.bidder.zoneId not provided",
+		}}
+	}
+
+	if smilewantedExt.ZoneID == "" {
+		return nil, []error{&errortypes.BadInput{
+			Message: "ZoneId is empty",
+		}}
+	}
+
+	validImpExists := false
+	for i := 0; i < len(request.Imp); i++ {
+		if request.Imp[i].Banner != nil {
+			bannerCopy := *requestIn.Imp[i].Banner
+			if len(bannerCopy.Format) > 0 {
+				firstFormat := bannerCopy.Format[0]
+				bannerCopy.W = &(firstFormat.W)
+				bannerCopy.H = &(firstFormat.H)
+				request.Imp[i].Banner = &bannerCopy
+				validImpExists = true
+			} else if bannerCopy.W != nil && bannerCopy.H != nil {
+				validImpExists = true
+			} else {
+				return nil, []error{&errortypes.BadInput{
+					Message: "No valid banner format in the bid request",
+				}}
+			}
+		} else if request.Imp[i].Video != nil {
+			validImpExists = true
+		}
+	}
+	if !validImpExists {
```

Comment:
The `validImpExists` check is not necessary. I see this is largely being used to ensure the impression is a banner and/or video, but that constraint is already present in `smilewanted.yaml` Prebid Server's code logic will check there are valid imps before calling your adapter. The banner format check still makes to keep.